File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/SyncBackendSetupActivity.java
package org.totschnig.myexpenses.activity;

import android.accounts.AccountManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.content.IntentSender;
import android.os.Bundle;
import android.view.Menu;
import android.view.SubMenu;

import com.annimon.stream.Exceptional;
import com.google.android.material.snackbar.Snackbar;

import org.jetbrains.annotations.NotNull;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.dialog.EditTextDialog;
import org.totschnig.myexpenses.dialog.NewMessageDialogFragment;
import org.totschnig.myexpenses.dialog.ProgressDialogFragment;
import org.totschnig.myexpenses.dialog.SetupWebdavDialogFragment;
import org.totschnig.myexpenses.model.ContribFeature;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.sync.ServiceLoader;
import org.totschnig.myexpenses.sync.SyncBackendProvider;
import org.totschnig.myexpenses.sync.SyncBackendProviderFactory;
import org.totschnig.myexpenses.sync.WebDavBackendProviderFactory;
import org.totschnig.myexpenses.task.SyncAccountTask;
import org.totschnig.myexpenses.task.TaskExecutionFragment;

import java.io.File;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import eltos.simpledialogfragment.form.Input;
import eltos.simpledialogfragment.form.SimpleFormDialog;
import eltos.simpledialogfragment.input.SimpleInputDialog;
import icepick.State;
import timber.log.Timber;

import static org.totschnig.myexpenses.activity.ConstantsKt.SYNC_BACKEND_SETUP_REQUEST;
import static org.totschnig.myexpenses.sync.GenericAccountService.KEY_PASSWORD_ENCRYPTION;
import static org.totschnig.myexpenses.sync.GenericAccountService.KEY_SYNC_PROVIDER_URL;
import static org.totschnig.myexpenses.sync.GenericAccountService.KEY_SYNC_PROVIDER_USERNAME;
import static org.totschnig.myexpenses.sync.WebDavBackendProvider.KEY_ALLOW_UNVERIFIED;
import static org.totschnig.myexpenses.sync.WebDavBackendProvider.KEY_WEB_DAV_CERTIFICATE;
import static org.totschnig.myexpenses.sync.WebDavBackendProvider.KEY_WEB_DAV_FALLBACK_TO_CLASS1;
import static org.totschnig.myexpenses.task.TaskExecutionFragment.TASK_CREATE_SYNC_ACCOUNT;
import static org.totschnig.myexpenses.task.TaskExecutionFragment.TASK_FETCH_SYNC_ACCOUNT_DATA;
import static org.totschnig.myexpenses.task.TaskExecutionFragment.TASK_WEBDAV_TEST_LOGIN;

public abstract class SyncBackendSetupActivity extends ProtectedFragmentActivity
    implements EditTextDialog.EditTextDialogListener, SimpleInputDialog.OnDialogResultListener {
  private static final String DIALOG_TAG_PASSWORD = "password";
  private static final int REQUEST_CODE_RESOLUTION = 1;

  protected List<SyncBackendProviderFactory> backendProviders;

  private boolean isResumed = false;
  private boolean setupPending = false;

  @State
  int selectedFactoryId;
  @NotNull
  public static final String KEY_SYNC_PROVIDER_ID  = "syncProviderId";

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    backendProviders = ServiceLoader.load(this);
  }

  //LocalFileBackend
  public void onFinishEditDialog(Bundle args) {
    String filePath = args.getString(EditTextDialog.KEY_RESULT);
    File baseFolder = new File(filePath);
    if (!baseFolder.isDirectory()) {
      showSnackbar("No directory " + filePath, Snackbar.LENGTH_SHORT);
    } else {
      String accountName = getSyncBackendProviderFactoryByIdOrThrow(R.id.SYNC_BACKEND_LOCAL).buildAccountName(filePath);
      Bundle bundle = new Bundle(1);
      bundle.putString(KEY_SYNC_PROVIDER_URL, filePath);
      createAccount(accountName, null, null, bundle);
    }
  }

  //WebDav
  public void onFinishWebDavSetup(Bundle data) {
    String userName = data.getString(AccountManager.KEY_ACCOUNT_NAME);
    String password = data.getString(AccountManager.KEY_PASSWORD);
    String url = data.getString(KEY_SYNC_PROVIDER_URL);
    String certificate = data.getString(KEY_WEB_DAV_CERTIFICATE);
    String accountName = getSyncBackendProviderFactoryByIdOrThrow(R.id.SYNC_BACKEND_WEBDAV).buildAccountName(url);

    Bundle bundle = new Bundle();
    bundle.putString(KEY_SYNC_PROVIDER_URL, url);
    bundle.putString(KEY_SYNC_PROVIDER_USERNAME, userName);
    if (certificate != null) {
      bundle.putString(KEY_WEB_DAV_CERTIFICATE, certificate);
    }
    if (data.getBoolean(KEY_WEB_DAV_FALLBACK_TO_CLASS1)) {
      bundle.putString(KEY_WEB_DAV_FALLBACK_TO_CLASS1, "1");
    }
    if (prefHandler.getBoolean(PrefKey.WEBDAV_ALLOW_UNVERIFIED_HOST, false)) {
      bundle.putString(KEY_ALLOW_UNVERIFIED, "true");
    }
    createAccount(accountName, password, null, bundle);
  }

  @Override
  protected void onResume() {
    super.onResume();
    isResumed = true;
    if (setupPending) {
      startSetupDo();
      setupPending = false;
    }
  }

  @Override
  protected void onPause() {
    super.onPause();
    isResumed = false;
  }

  public void startSetup(int itemId) {
    selectedFactoryId = itemId;
    if (isResumed) {
      startSetupDo();
    } else {
      setupPending = true;
    }
  }

  private void startSetupDo() {
    SyncBackendProviderFactory syncBackendProviderFactory =
        getSyncBackendProviderFactoryById(selectedFactoryId);
    if (syncBackendProviderFactory != null) {
      syncBackendProviderFactory.startSetup(this);
    }
  }

  //Google Drive & Dropbox
  @Override
  protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
    super.onActivityResult(requestCode, resultCode, intent);
    if (requestCode == SYNC_BACKEND_SETUP_REQUEST && resultCode == RESULT_OK && intent != null) {
      String accountName = getSyncBackendProviderFactoryByIdOrThrow(intent.getIntExtra(KEY_SYNC_PROVIDER_ID, 0))
          .buildAccountName(intent.getStringExtra(AccountManager.KEY_ACCOUNT_NAME));
      createAccount(accountName, null,
          intent.getStringExtra(AccountManager.KEY_AUTHTOKEN), intent.getBundleExtra(AccountManager.KEY_USERDATA));
    }
    if (requestCode == REQUEST_CODE_RESOLUTION) {
      showSnackbar("Please try again");
    }
  }

  protected void createAccount(String accountName, String password, String authToken, Bundle bundle) {
    Bundle args = new Bundle();
    args.putString(AccountManager.KEY_ACCOUNT_NAME, accountName);
    args.putString(AccountManager.KEY_PASSWORD, password);
    args.putString(AccountManager.KEY_AUTHTOKEN, authToken);
    args.putParcelable(AccountManager.KEY_USERDATA, bundle);
    args.putBoolean(SyncAccountTask.KEY_RETURN_REMOTE_DATA_LIST, createAccountTaskShouldReturnDataList());
    SimpleFormDialog.build().msg(R.string.passphrase_for_synchronization)
        .fields(Input.password(KEY_PASSWORD_ENCRYPTION).required().hint(R.string.input_label_passphrase))
        .extra(args)
        .neut(R.string.button_label_no_encryption)
        .show(this, DIALOG_TAG_PASSWORD);
  }

    protected void createAccountDo(Bundle args) {
    getSupportFragmentManager()
        .beginTransaction()
        .add(TaskExecutionFragment.newInstanceWithBundle(args, TASK_CREATE_SYNC_ACCOUNT), ASYNC_TAG)
        .add(ProgressDialogFragment.newInstance(getString(R.string.progress_dialog_fetching_data_from_sync_backend)), PROGRESS_TAG)
        .commit();
  }

  public void fetchAccountData(String accountName) {
    Bundle args = new Bundle();
    args.putString(AccountManager.KEY_ACCOUNT_NAME, accountName);
    args.putBoolean(SyncAccountTask.KEY_RETURN_REMOTE_DATA_LIST, true);
    getSupportFragmentManager()
        .beginTransaction()
        .add(TaskExecutionFragment.newInstanceWithBundle(args, TASK_FETCH_SYNC_ACCOUNT_DATA), ASYNC_TAG)
        .add(ProgressDialogFragment.newInstance(getString(R.string.progress_dialog_fetching_data_from_sync_backend)), PROGRESS_TAG)
        .commit();
  }

  protected boolean createAccountTaskShouldReturnDataList() {
    return false;
  }

  public void onCancelEditDialog() {

  }

  @Override
  public void onPostExecute(int taskId, Object o) {
    super.onPostExecute(taskId, o);
    switch (taskId) {
      case TASK_CREATE_SYNC_ACCOUNT: {
        final Exceptional exceptional = (Exceptional) o;
        if (exceptional.isPresent()) {
          recordUsage(ContribFeature.SYNCHRONIZATION);
          if ("xiaomi".equalsIgnoreCase(android.os.Build.MANUFACTURER)) {
            NewMessageDialogFragment.Companion.newInstance("On some Xiaomi devices, synchronization does not work without Autostart permission. Visit <a href=\"https://github.com/mtotschnig/MyExpenses/wiki/FAQ:-Synchronization#q2\">MyExpenses FAQ</a> for more information.", true).show(getSupportFragmentManager(), "XIAOMI_WARNING");
          }
        } else {
          Throwable throwable = exceptional.getException();
          if (throwable instanceof SyncBackendProvider.ResolvableSetupException) {
            try {
              final PendingIntent resolution = ((SyncBackendProvider.ResolvableSetupException) throwable).getResolution();
              if (resolution != null) {
                startIntentSenderForResult(resolution.getIntentSender(), REQUEST_CODE_RESOLUTION, null, 0, 0, 0);
              }
            } catch (IntentSender.SendIntentException e) {
              Timber.e(e, "Exception while starting resolution activity");
            }
          } else {
            showSnackbar("Unable to set up account: " + throwable.getMessage());
          }
        }
        break;
      }
      case TASK_WEBDAV_TEST_LOGIN: {
        getWebdavFragment().onTestLoginResult((Exceptional<Void>) o);
        break;
      }
    }
  }

  public void addSyncProviderMenuEntries(SubMenu subMenu) {
    for (SyncBackendProviderFactory factory : backendProviders) {
      subMenu.add(Menu.NONE, factory.getId(), Menu.NONE, factory.getLabel());
    }
  }

  public @Nullable
  SyncBackendProviderFactory getSyncBackendProviderFactoryById(int id) {
    try {
      return getSyncBackendProviderFactoryByIdOrThrow(id);
    } catch (IllegalStateException e) {
      return null;
    }
  }

  public @NonNull
  SyncBackendProviderFactory getSyncBackendProviderFactoryByIdOrThrow(int id) throws IllegalStateException {
    for (SyncBackendProviderFactory factory : backendProviders) {
      if (factory.getId() == id) {
        return factory;
      }
    }
    throw new IllegalStateException();
  }

  protected SetupWebdavDialogFragment getWebdavFragment() {
    return (SetupWebdavDialogFragment) getSupportFragmentManager().findFragmentByTag(
        WebDavBackendProviderFactory.WEBDAV_SETUP);
  }

  @Override
  public boolean onResult(@NonNull String dialogTag, int which, @NonNull Bundle extras) {
    if (DIALOG_TAG_PASSWORD.equals(dialogTag)) {
      if (which != BUTTON_POSITIVE || "".equals(extras.getString(KEY_PASSWORD_ENCRYPTION))) {
        extras.remove(KEY_PASSWORD_ENCRYPTION);
      }
      createAccountDo(extras);
    }
    return false;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/SyncBackendSetupActivity.kt
package org.totschnig.myexpenses.activity

import android.accounts.AccountManager
import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.core.app.NotificationManagerCompat
import com.evernote.android.state.State
import eltos.simpledialogfragment.SimpleDialog.OnDialogResultListener
import eltos.simpledialogfragment.form.Input
import eltos.simpledialogfragment.form.SimpleFormDialog
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.feature.Feature
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.model.ContribFeature
import org.totschnig.myexpenses.sync.BackendService
import org.totschnig.myexpenses.sync.BackendService.Companion.allAvailable
import org.totschnig.myexpenses.sync.GenericAccountService
import org.totschnig.myexpenses.sync.json.AccountMetaData
import org.totschnig.myexpenses.util.PermissionHelper
import org.totschnig.myexpenses.util.safeMessage
import org.totschnig.myexpenses.viewmodel.SyncViewModel

abstract class SyncBackendSetupActivity : RestoreActivity(),
    OnDialogResultListener {

    private lateinit var backendProviders: List<BackendService>
    protected val syncViewModel: SyncViewModel by viewModels()
    private var isResumed = false

    @State
    var selectedFactoryId = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        backendProviders = BackendService.allAvailable(this)
        injector.inject(syncViewModel)
    }

    override fun onResume() {
        super.onResume()
        isResumed = true
        if (selectedFactoryId != 0) {
            startSetupCheckPrerequisites()
        }
    }

    override fun onPause() {
        super.onPause()
        isResumed = false
    }

    fun startSetup(itemId: Int) {
        selectedFactoryId = itemId
        if (isResumed) {
            startSetupCheckPrerequisites()
        }
    }

    private val startSetup =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                result.data?.let {
                    createAccount(
                        it.getStringExtra(AccountManager.KEY_ACCOUNT_NAME)!!,
                        it.getStringExtra(AccountManager.KEY_PASSWORD),
                        it.getStringExtra(AccountManager.KEY_AUTHTOKEN),
                        it.getBundleExtra(AccountManager.KEY_USERDATA)
                    )
                }
            }
        }

    private fun startSetupCheckPrerequisites() {
        val backendService = getBackendServiceById(selectedFactoryId)
        val feature = backendService?.feature
        if (feature == null || featureManager.isFeatureInstalled(feature, this)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                !NotificationManagerCompat.from(this).areNotificationsEnabled()
            ) {
                requestNotificationPermission(PermissionHelper.PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC)
            } else {
                startSetupDo()
            }
        } else {
            featureManager.requestFeature(feature, this)
        }
    }

    private fun startSetupDo() {
        getBackendServiceById(selectedFactoryId)
            ?.instantiate()
            ?.getOrNull()
            ?.setupActivityClass
            ?.let {
                startSetup.launch(Intent(this, it))
            }
        selectedFactoryId = 0
    }

    override fun onFeatureAvailable(feature: Feature) {
        super.onFeatureAvailable(feature)
        if (selectedFactoryId != 0 && getBackendServiceByIdOrThrow(selectedFactoryId).feature == feature) {
            startSetupCheckPrerequisites()
        }
    }

    override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
        super.onPermissionsDenied(requestCode, perms)
        if (requestCode == PermissionHelper.PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC) {
            startSetupDo()
        }
    }

    override fun onPermissionsGranted(requestCode: Int, perms: List<String>) {
        super.onPermissionsGranted(requestCode, perms)
        if (requestCode == PermissionHelper.PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC) {
            startSetupDo()
        }
    }

    private fun createAccount(
        accountName: String?,
        password: String?,
        authToken: String?,
        bundle: Bundle?
    ) {
        val args = Bundle().apply {
            putString(AccountManager.KEY_ACCOUNT_NAME, accountName)
            putString(AccountManager.KEY_PASSWORD, password)
            putString(AccountManager.KEY_AUTHTOKEN, authToken)
            putParcelable(AccountManager.KEY_USERDATA, bundle)
        }
        if (offerEncryption) {
            SimpleFormDialog
                .build()
                .msg(R.string.passphrase_for_synchronization)
                .fields(
                    Input
                        .password(GenericAccountService.KEY_PASSWORD_ENCRYPTION)
                        .required()
                        .hint(R.string.input_label_passphrase)
                )
                .extra(args)
                .neut(R.string.button_label_no_encryption)
                .show(this, DIALOG_TAG_PASSWORD)
        } else {
            createAccountDo(args)
        }
    }

    private fun createAccountDo(args: Bundle) {
        showLoadingSnackBar()
        syncViewModel.createSyncAccount(
            args,
            createAccountTaskShouldQueryLocalAccounts,
            createAccountTaskShouldReturnBackups,
            createAccountTaskShouldQueryRemoteAccounts
        ).observe(this) { result ->
            result.onSuccess {
                dismissSnackBar()
                recordUsage(ContribFeature.SYNCHRONIZATION)
                if ("xiaomi".equals(Build.MANUFACTURER, ignoreCase = true)) {
                    showMessage("On some Xiaomi devices, synchronization does not work without AutoStart permission. Visit <a href=\"https://faq.myexpenses.mobi/sync-xiaomi\">MyExpenses FAQ</a> for more information.")
                }
                onReceiveSyncAccountData(it)
            }.onFailure { throwable ->
                showSnackBar("Unable to set up account: " + throwable.message)
            }
        }
    }

    abstract fun onReceiveSyncAccountData(data: SyncViewModel.SyncAccountData)

    fun checkForDuplicateUuids(data: List<AccountMetaData>) =
        data.map(AccountMetaData::uuid).distinct().count() < data.count()

    fun fetchAccountData(accountName: String) {
        showLoadingSnackBar()
        syncViewModel.fetchAccountData(accountName).observe(this) { result ->
            dismissSnackBar()
            result.onSuccess {
                onReceiveSyncAccountData(it)
            }.onFailure {
                showSnackBar(it.safeMessage)
            }
        }
    }

    fun showLoadingSnackBar() {
        showProgressSnackBar(getString(R.string.progress_dialog_fetching_data_from_sync_backend))
    }

    protected open val createAccountTaskShouldReturnBackups = false

    protected open val createAccountTaskShouldQueryLocalAccounts = true

    protected open val createAccountTaskShouldQueryRemoteAccounts = true

    /**
     * When the backend is setup from Backup / Restore preference screen, the account specific encryption
     * is not relevant, since encryption of backups is configured by a global setting
     */
    protected open val offerEncryption = true

    fun addSyncProviderMenuEntries(subMenu: Menu) {
        for (factory in allAvailable(this)) {
            subMenu.add(Menu.NONE, factory.id, Menu.NONE, factory.label)
        }
    }

    fun getBackendServiceById(id: Int) = try {
        getBackendServiceByIdOrThrow(id)
    } catch (e: NoSuchElementException) {
        null
    }

    @Throws(NoSuchElementException::class)
    fun getBackendServiceByIdOrThrow(id: Int) = allAvailable(this).first { it.id == id }

    override fun onResult(dialogTag: String, which: Int, extras: Bundle) =
        when {
            super.onResult(dialogTag, which, extras) -> true
            DIALOG_TAG_PASSWORD == dialogTag -> {
                if (which != OnDialogResultListener.BUTTON_POSITIVE || "" == extras.getString(
                        GenericAccountService.KEY_PASSWORD_ENCRYPTION
                    )
                ) {
                    extras.remove(GenericAccountService.KEY_PASSWORD_ENCRYPTION)
                }
                createAccountDo(extras)
                true
            }

            else -> false
        }

    companion object {
        private const val DIALOG_TAG_PASSWORD = "password"
        const val REQUEST_CODE_RESOLUTION = 1
    }
}